MagneticField/VolumeBasedEngine/src/MagGeometry.cc

```diff
@@ -148,15 +176,17 @@ MagVolume const* MagGeometry::findVolume1(const GlobalPoint& gp, double toleranc
 
 // Use hierarchical structure for fast lookup.
 MagVolume const* MagGeometry::findVolume(const GlobalPoint& gp, double tolerance) const {
-  // Check volume cache
-  auto lastVolumeCheck = lastVolume.load(std::memory_order_acquire);
-  if (lastVolumeCheck != nullptr && lastVolumeCheck->inside(gp)) {
-    return lastVolumeCheck;
-  }
+  static thread_local MagVolume const* lastVolume = nullptr;
+
+  MagVolume const* result = lastVolume;
 
-  MagVolume const* result = nullptr;
-  if (inBarrel(gp)) {  // Barrel
-    double R = gp.perp();
+  if (result && result->inside(gp))
+    return result;
```

Comment:
generally, MagGeometry is not a singleton. So, using a static here will lead to incorrect behavior for the odd cases where MagGeometry is not unique.
It's perhaps safe enough to add a configurable to choose between atomic and thread_local way of caching.

@Dr15Jones @makortel is there another way to make ES products have per-stream/thread caches?

Reply:
> is there another way to make ES products have per-stream/thread caches?

There is not. Please avoid thread_local as well since we've had problems with some systems based on the the total number of thread_local being used in an application.

Reply:
If one wants to sped up searches, returning the last hit volume as part of the API and having the caller pass it back it is the best thread-safe way to do it.

Reply:
@slava77, you are right that MagGeometry is not guaranteed to be unique, but we don't have real use cases where two instances would be present. At least it should not happen in any of the standard configurations, right?
@Dr15Jones: the first problem with returning the volume to the API is that volume search is a detail of the implementation of this specific field provider. Volume caching is an internal optimization specific of this provider. The general MF interface does not know about volumes and we have maps that are not volume-based; in particular we also serve a fast simplified parametrization of the tracker that is not based on volumes, cf. https://github.com/cms-sw/cmssw/blob/CMSSW_10_3_X/Configuration/StandardSequences/python/MagneticField_cff.py
The second problem that I see is that having to pass back the pointer and let the caller handle it will require substantial modification in CMSSW, and I can imagine that in some cases it will be hard to reuse the pointer without some refactoring (imagine a call to the field within a class method, or if the field is called in several parts of some functional operation)  so that practically it will lead to dropping this optimization in many cases.

Reply:
I agree with both @Dr15Jones and @namapane that the best solution would be to let the caller handle the caches, but that requires some extra refactoring throughout other parts of the code. I made my changes considering that, even if not a singleton, realistically only a single instance of the class exists. The thread local cache does have a cost, but is probably a good compromise given the circumstances. Note that the main magnetic field has the same problem and we need to fix it as well. See, e.g., https://github.com/cms-sw/cmssw/blob/03d4c04f6c7c596ddad1a8d4193d322a7dcd9e0c/SimG4Core/MagneticField/interface/Field.h#L19-L20 and https://github.com/cms-sw/cmssw/blob/master/SimG4Core/MagneticField/src/Field.cc#L20-L39